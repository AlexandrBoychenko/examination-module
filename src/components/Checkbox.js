import React from 'react';
import { setLocalStorage, returnCurrentValue } from '../helpers'

class Checkbox extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            checked: []
        };
        this.handleChangeCheckbox = this.handleChangeCheckbox.bind(this);
    }

    componentDidMount() {
        setLocalStorage(this.props.id, this.props.right);
        this.compareWithState();
    }

    handleChangeCheckbox(e) {
        let currentValues = this.state.checked.slice();
        let targetPos = currentValues.indexOf(e.target.value);
        if (~targetPos) {
            currentValues.splice(targetPos, 1);
        } else {
            currentValues.push(e.target.value);
        }
        this.setState({checked: currentValues});
        this.props.context.onAnswerChangeCheckbox(e.target.value, this.props.id, 'Checkbox');
    }

    renderItems() {
        return this.props.items.map((item) => {
            return this.renderInput(item)
        });
    }

    compareWithState() {
        let arrayForChecked = [];
        let pastValues = this.props.pastValues;
        this.props.items.forEach((item) => {
            if (pastValues && pastValues[this.props.id]) {
                pastValues[this.props.id].forEach((propsItem) => {
                    if(propsItem === item) {
                        arrayForChecked.push(item.toString());
                    }
                })
            }
        });
        this.setState({checked: arrayForChecked});
    }

    isItemChecked(item) {
        return ~this.state.checked.indexOf(item.toString())
    }

    renderInput(item) {
        this.className = 'checkbox-list';
        return (
            <li key={item}>
                <label className="container container-checkbox">{item}
                    <input type="checkbox" name="answer" value={item}
                       onChange={this.handleChangeCheckbox}
                       checked={returnCurrentValue(this.props.parentState.length, this.isItemChecked(item))}
                    />
                    <span className="check-mark-checkbox"></span>
                </label>
            </li>
        )
    }

    render() {
        return (
            <ul className={this.className}>
                {this.renderItems()}
            </ul>
        )
    }
}

export default Checkbox;
