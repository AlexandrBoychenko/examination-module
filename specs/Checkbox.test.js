import React from 'react';
import Checkbox from '../src/Components/Checkbox';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';


describe('Checkbox', () => {
    const testCheckboxProps = {
        id: "question02",
        items: [
            "Имеет хвост",
            "Вращается вокруг солнца",
            "Состоит из газа и пыли",
            "Существует только во внутренней области солнечной системы",
            "Не имеет ядра"
        ],
        right: [
            "Имеет хвост",
            "Вращается вокруг солнца",
        ],
        context: {onAnswerChangeCheckbox: () => {}},
        pastValues: {},
        parentState: {}
    };

    const onChange = jest.fn(),
        props = {
            ...testCheckboxProps,
            onChange
        };
    const wrapper = mount(<Checkbox {...props} />);
    const CheckboxInputComponent = wrapper.find("input").first();

    test('It should be render correctly', () => {
        const component = renderer.create(<Checkbox {...testCheckboxProps} />);
        let treeHTML = component.toJSON();
        expect(treeHTML).toMatchSnapshot();
    });

    test('it should have correct props types for each input', () => {
        const wrapper = mount(<Checkbox {...testCheckboxProps} />);
        const CheckboxProps = wrapper.find("input").first().props();

        expect(CheckboxProps.type).toEqual('checkbox');
        expect(CheckboxProps.name).toEqual(wrapper.props().id);
        expect(CheckboxProps.value).toEqual(expect.any(String));
        expect(CheckboxProps.checked).toEqual("");
        expect(CheckboxProps.onChange).toEqual(expect.any(Function));
    });

    test("check the onChange callback", () => {
        CheckboxInputComponent.simulate("change", {target: {"value": props.items[0]}});
        expect(wrapper.state().checked).toEqual([CheckboxInputComponent.props().value]);
    });

    test("check the state after dual onChange callback", () => {
        CheckboxInputComponent.simulate("change", {target: {"value": props.items[0]}});
        CheckboxInputComponent.simulate("change", {target: {"value": props.items[3]}});
        CheckboxInputComponent.simulate("change", {target: {"value": ""}});
        expect(wrapper.state().checked[0]).toEqual(props.items[3]);
    });
});
