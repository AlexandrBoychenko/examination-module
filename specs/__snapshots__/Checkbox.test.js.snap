// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox It should be render correctly 1`] = `
<ul>
  <li>
    <label
      className="container container-checkbox"
    >
      Имеет хвост
      <input
        checked=""
        name="question02"
        onChange={[Function]}
        type="checkbox"
        value="Имеет хвост"
      />
      <span
        className="check-mark-checkbox"
      />
    </label>
  </li>
  <li>
    <label
      className="container container-checkbox"
    >
      Вращается вокруг солнца
      <input
        checked=""
        name="question02"
        onChange={[Function]}
        type="checkbox"
        value="Вращается вокруг солнца"
      />
      <span
        className="check-mark-checkbox"
      />
    </label>
  </li>
  <li>
    <label
      className="container container-checkbox"
    >
      Состоит из газа и пыли
      <input
        checked=""
        name="question02"
        onChange={[Function]}
        type="checkbox"
        value="Состоит из газа и пыли"
      />
      <span
        className="check-mark-checkbox"
      />
    </label>
  </li>
  <li>
    <label
      className="container container-checkbox"
    >
      Существует только во внутренней области солнечной системы
      <input
        checked=""
        name="question02"
        onChange={[Function]}
        type="checkbox"
        value="Существует только во внутренней области солнечной системы"
      />
      <span
        className="check-mark-checkbox"
      />
    </label>
  </li>
  <li>
    <label
      className="container container-checkbox"
    >
      Не имеет ядра
      <input
        checked=""
        name="question02"
        onChange={[Function]}
        type="checkbox"
        value="Не имеет ядра"
      />
      <span
        className="check-mark-checkbox"
      />
    </label>
  </li>
</ul>
`;
